src/shogun/machine/LinearMachine.h

```diff
@@ -142,6 +134,18 @@ class LinearMachine : public Machine
 		 */
 		virtual SGVector<float64_t> apply_get_outputs(std::shared_ptr<Features> data);
 
+		bool train_machine(const std::shared_ptr<Features>& data, const std::shared_ptr<Labels>& labs) final
+		{
+			const auto dot_feat = std::dynamic_pointer_cast<DotFeatures>(data);
+			require(dot_feat, "Expected DotFeatures type");
+			return train_machine(dot_feat, labs);
+		}
+
+		virtual bool train_machine(const std::shared_ptr<DotFeatures>& data, const std::shared_ptr<Labels>& labs)
```

Comment:
can it not be pure virtual?

Reply:
some place need linearmachine instance.

Reply:
You mean they instantiate `LinearMachine`?

Reply:
yes

Reply:
@karlnapf there are some really weird things going on in some classes where we instantiate LinearMachine, even though I am pretty sure this class should be abstract

Reply:
eh? how? that should not be the case.
@LiuYuHui could you point me to the line where std::make_shared<LinearMachine> is being called?

Reply:
my favourite https://github.com/shogun-toolbox/shogun/blob/aede117e19ec3530583261ce3ca21879a2612495/src/shogun/multiclass/MulticlassLibLinear.cpp#L132